crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2830,95 +2567,387 @@ impl<'db> Type<'db> {
-                ClassBase::Dynamic(dynamic_type) => {
-                    Type::Dynamic(dynamic_type).try_call(db, arguments)
-                }
-                ClassBase::Class(class) => Type::class_literal(class).try_call(db, arguments),
+                ClassBase::Dynamic(dynamic_type) => Type::Dynamic(dynamic_type)
+                    .signatures(db, callable_ty)
+                    .clone(),
+                ClassBase::Class(class) => Type::class_literal(class)
+                    .signatures(db, callable_ty)
+                    .clone(),
             },
 
-            instance_ty @ Type::Instance(_) => {
-                instance_ty
-                    .try_call_dunder(db, "__call__", arguments)
-                    .map_err(|err| match err {
-                        CallDunderError::Call(CallError::NotCallable { .. }) => {
-                            // Turn "`<type of illegal '__call__'>` not callable" into
-                            // "`X` not callable"
-                            CallError::NotCallable {
-                                not_callable_type: self,
-                            }
+            Type::Instance(_) => {
+                // Note that for objects that are callable via a `__call__` method, we will get the
+                // signature of the dunder method, but will pass in the type of the object as the
+                // "callable type". That ensures that we get errors like "`X` is not callable"
+                // instead of "`<type of illegal '__call__'>` is not callable".
+                let Some((signatures, boundness)) =
+                    self.dunder_signature(db, Some(callable_ty), "__call__")
+                else {
+                    return Signatures::not_callable(callable_ty, self);
+                };
+                let mut signatures = signatures.clone();
+                signatures.set_dunder_call_boundness(boundness);
+                signatures
+            }
+
+            // Dynamic types are callable, and the return type is the same dynamic type. Similarly,
+            // `Never` is always callable and returns `Never`.
+            Type::Dynamic(_) | Type::Never => Signatures::single(CallableSignature::dynamic(self)),
+
+            // Note that this correctly returns `None` if none of the union elements are callable.
+            Type::Union(union) => Signatures::from_union(
+                callable_ty,
+                self,
+                union
+                    .elements(db)
+                    .iter()
+                    .map(|element| element.signatures(db, *element).clone()),
+            ),
+
+            Type::Intersection(_) => {
+                Signatures::single(CallableSignature::todo("Type::Intersection.call()"))
+            }
+
+            _ => Signatures::not_callable(callable_ty, self),
+        }
+    }
+
+    /// Calls `self`
+    fn call(self, db: &'db dyn Db, arguments: &CallArguments<'_, 'db>) -> Bindings<'db> {
```

Comment:
I overall like the change, but I'm a bit concerned about `call` returning `Bindings` directly and relying on the caller's *good behavior* to only call `bindings.return_type` if they're okay ignoring any possible binding error. It somewhat undoes what I tried to accomplish with my `try_` refactor to make all error handling explicit. 

The method name also doesn't fit into the `try_` for faliable operations naming-convention anymore. It should also simplify the call-sites because every call site immediately calls `as_result`. 

I very much like the outcome that bindings now stores the relevant information.
A possible solution here is to change `CallError` to a struct with two fields: `Bindings` and `kind` where kind is what is `CallError` in your PR.

Reply:
Yeah I was going back on forth on that.  I had thought that some callers would want to inspect the bindings before converting to a result.  But no one is doing that, and even if they did, I can probably add a helper method to make that easier.

If you look a couple of commits back that's sort of how it was...though the lower layer bindings need to return a `CallError` without the attached `Bindings`, so I had it with a parameterized type...  Anyway I think I can pull out a `CallErrorKind` to make this work.  Will try that now

Reply:
Done